### PR TITLE
[cache] Delayed resolution of factories and listener configurations for caches

### LINKED ISSUE #11905: [cache] Implement delayed resolution for other types in CacheConfig
In addition to key and value types, the following need to be deferred:
```cacheLoaderFactory, cacheWriterFactory, expiryPolicyFactory and listenerConfigurations```
The above also rely for their classes to exist in current Class Loader and may not,
causing the node initialization failures.

There should be a similar fix for these as in key and value types.
Suggestion would be to transfer an array of bytes rather than objects for these values,
and actually de-serialize them only when needed.

Base issue: #8972 and PR https://github.com/hazelcast/hazelcast/pull/10874

Related Issue #10728
Related Pull Request https://github.com/hazelcast/hazelcast/pull/11513

### PR DESCRIPTION
**Pull request by lprimak**
just like key / value types in case classes are not available at
pre-join time
fixes #11905

### CODE REVIEW

**Comment by lprimak** [comment]
Needs to be coordinate with #10874

**Comment by lprimak** [comment]
This needs a little bit of work because some tests are failing

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by lprimak** [comment]
test suite is working now, will add own test

**Comment by lprimak** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
@lprimak in principle, looks good. We should be able to obtain and use the actual `SerializationService` of the Hazelcast member, so that any user configuration is reflected in the serialization service output. Also, maybe it makes sense to use `Data` instead of `byte[]` as an opaque data container. I will take a better look at these two points and update you.

In the meanwhile it would be great if you can add a few tests in `PreJoinCacheConfigTest` that verify the added code (ie a non-resolved `PreJoinCacheConfig`'s factories are properly returned from the relevant getters etc). I also have some more comments (eg in general we don't use underscores to prefix field names), but we can get to those once we get the functionality right.

**Comment by lprimak** [comment]
I have tried to convert it to Data, but it doesn't fit the situation because writeListenerConfigurations() writes to a stream, and I want to reuse that code, which I don't think is compatible with the Data's API model, as toData() just takes one object

**Comment by lprimak** [comment]
Figured out a way to get the serialization service, just like ClassLoader is retrieved

**Comment by lprimak** [comment]
Ok, I am basically done with all your comments above

File: hazelcast/src/main/java/com/hazelcast/cache/impl/JCacheDetector.java [hunk 1]
```diff
@@ -45,7 +45,9 @@ private JCacheDetector() {
     }
 
     public static boolean isJCacheAvailable(ClassLoader classLoader) {
-        return isJCacheAvailable(classLoader, null);
+        ClassLoader backupClassLoader = JCacheDetector.class.getClassLoader();
```
[the remark this answers is not in the document]
Comment (lprimak) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/JCacheDetector.java]: Fixed

File: hazelcast/src/main/java/com/hazelcast/config/AbstractCacheConfig.java [hunk 2]
```diff
@@ -423,14 +436,59 @@ public boolean isStoreByValue() {
...
+
+    private void resolveDelayedLoadingClasses() {
+        if(serializedFactories != null) {
+            if (classLoader != null) {
+                Thread.currentThread().setContextClassLoader(classLoader);
+            }
+            // TODO set tenant so the classes resolve
+            ObjectDataInputStream factoriesStrm = null;
+            ObjectDataInputStream listenersStrm = null;
+            ClassLoader oldThrClassLoader = Thread.currentThread().getContextClassLoader();
```
[the remark this answers is not in the document]
Comment (lprimak) [reply hazelcast/src/main/java/com/hazelcast/config/AbstractCacheConfig.java]: This is correct, and TCCL should further be overridden if the tenant so desires, becase the classLoader field is the config field, which may not be enough for the tenant.

I have fixed the comments and the old class loader save was happening too late

**Comment by lprimak** [comment]
superceded by #12624

**Pull request closed without merging**